add calc_growthrate function

orichters opened the pull request:
Calculates the growth rate of all variables in '%/yr' and returns a sorted data.frame where each variable name gets ` [Growth rate]` appended.

If, for example, your data contains the data in 2070 and 2060, the growth rate returned for 2070 is calculated as 100 * ((d2070/d2060)^(1/10) - 1). No growth rate can be calculated for the first year of the data. Infinite or undefined values (for example if d2060 = 0) are dropped.

Is the name ok or is there a risk of conflict with functions from other packages?

pweigmann commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+    group_by(.data$model, .data$scenario, .data$region, .data$variable) %>%
+    mutate(diffyear = .data$period - lag(.data$period)) %>%
+    mutate(growthrate = 100 * ((.data$value/lag(.data$value))^(1/.data$diffyear) - 1)) %>%
+    filter(! .data$growthrate %in% c(NA, Inf, NaN)) %>%
```
just in case also include `-Inf` maybe?

orichters replied:
good idea

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
How about you figure out when you are done and request reviews then, so people do not have to play whack-a-mole …

pweigmann commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+    as.quitte(na.rm = TRUE) %>%
+    quitteSort() %>%
+    group_by(.data$model, .data$scenario, .data$region, .data$variable) %>%
+    mutate(diffyear = .data$period - lag(.data$period)) %>%
```
Exists in multiple packages, better use via ``package::lag``?

orichters replied:
I added it to the imports

pweigmann commented:
the function is quite slow because of ``mutate(variable = factor(paste0(.data$variable, " [Growth Rate]")))``, maybe operate on levels later in the function

orichters commented:
> How about you figure out when you are done and request reviews then, so people do not have to play whack-a-mole …

I agree, sorry, using the new IAMC style was an idea that came to my mind only later.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+#' Calculates the growth rate of all variables in '%/yr' and returns a sorted data.frame
+#' where each variable name gets '|Growth rate' appended.
+#' If, for example, your data contains the data in 2070 and 2060, the growth rate returned
+#' for 2070 is calculated as 100 * ((d2070/d2060)^(1/10) - 1).
+#' No growth rate can be calculated for the first year of the data.
+#' Infinite or undefined values (for example if d2060 = 0) are dropped.
```
Bit of a long title.


![image](https://github.com/pik-piam/quitte/assets/53254462/580902e5-7235-4ae8-a959-51516f3cf640)

orichters replied:
Moved it into "details"

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+#' No growth rate can be calculated for the first year of the data.
+#' Infinite or undefined values (for example if d2060 = 0) are dropped.
+#'
+#' @param x A quitte object
```
Actually takes anything that has an `as.quitte()` method.  It is an undocumented feature people might come to rely on.

orichters replied:
adjusted it, also in quitteSort

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+#'
+#' @return the sorted quitte object
+#' @export
+growthrate <- function(x) {
```
I would prefer are verb or verb & noun name for a function that transforms data, so e.g. `calc_growthrate()`.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+  x <- as.quitte(x, na.rm = TRUE)
+  levels(x$variable) <- paste(levels(x$variable), "[Growth Rate]")
+  x %>%
+    quitteSort() %>%
```
Why is the data frame sorted?  To ensure that `lag()` uses the previous period?  That should be done explicitly using `lag(order_by = .data$period)`.

orichters replied:
thanks, I adjusted it

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+    mutate(diffyear = .data$period - lag(.data$period),
+           growthrate = 100 * ((.data$value/lag(.data$value))^(1/.data$diffyear) - 1),
+           unit = factor("%/yr")) %>%
+    filter(! .data$growthrate %in% c(NA, -Inf, Inf, NaN)) %>%
```
`filter(is.finite(.data$growthrate))`

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented on R/growthrate.R:
```diff
@@ -0,0 +1,29 @@
+#' Calculates the growth rate of all variables in '%/yr' and returns a sorted data.frame
+#' where each variable name gets '|Growth rate' appended.
```
> gets '|Growth rate' appended

That part is not true anymore.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q reviewed (commented):
While I understand this function is designed to fit the task at hand, I feel it would be easy to extend to make more useful for other applications.
Specifically, people might want to calculate growth rates only for a subset of variables and add them to their data frame for further processing.
They would have to do some 
```
bind_rows(
    x,

    x %>%
        filter(…) %>%
        calc_growthrate()
)
```
which could easily be wrapped in this function, with arguments for filtering variables and appending instead of replacing.

orichters commented on R/calc_growthrate.R:
```diff
@@ -0,0 +1,30 @@
+#' Calculates the growth rate of all variables in '%/yr' and returns a quitte data.frame
```
I shortened it further

orichters commented:
I implemented `only.new` as in `calc_addVariable` and `filter.function` as in `read.quitte`, with the special case that you can also simply pass a vector `c("GDP|PPP")` and it will be then use only those. Added some more tests and found some minor bugs (missing `ungroup`), but should be good to go now.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q reviewed (approved):
Nice.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
This would actually be a good candidate to mention in the tutorial, @orichters.

orichters commented:
quitte or remindmodel tutorial?
